### PR TITLE
remove dead dependency: single

### PR DESCRIPTION
**Pull request by alex-ozdemir**
it depends on failure, which is deprecated

### CODE REVIEW

**Comment by alex-ozdemir** [comment]
Paging @CAD97. Not sure who maintains this crate, if anyone?

**Comment by CAD97** [comment]
I'm officially the maintainer of both pest-ast and single.

I consider single deprecated; itertools offers a better version of the functionality now.

Pest-ast, unfortunately, while it's functional and I know some people are (reasonably) happy with it, never really got past the experiment phase. The typed view it provided was incorporated into pest 3 plans, effort was moved to that project, and then that fell through.

I'm also not super happy with pest-ast: defining a PEG parser[^1] declaratively and then your AST declaratively separately does not sit well. Despite parsing and the (high level) AST being different problems, they share a significant amount of structure. But due to how writing PEG works, you end up with a parse tree (and “grammar” definition) significantly different from the logical syntax tree (even before any abstracting layers) corrupted by implementation details; solutions to remove left recursion, flat binop chains[^2] fixed by a later pratt parsing pass, and such.

[^1]: Somewhat controversial opinion: a PEG “grammar,” while it does define a grammar in the formal sense (i.e. a set of strings), as far as authoring and using it goes, is much better described as a declarative way to do recursive descent parsing. PEG is literally “Parsing Expression Grammar,” so this should be obvious, but it's very much not what people expect when they see a “parser generator” that generates a parser based on a “grammar.”

[^2]: This is emblematic of PEG being a parser grammar first: how you do solve left recursion is exactly how left recursion is solved in recursive descent: as a simple `expr ~ (op ~ expr)*` loop.

Add onto this other trivial things pest makes unnecessarily difficult; things like disallowing keyword as identifiers is trivially implicit[^3] when using a separate lexical classification phase, but pest uses a lexerless design. (Which I will admit is still very much in vogue; even generators which use split lexing are most commonly used with “lexerless” definitions mixing lexer and parser rules... which makes explaining the difference in behavior in multipass and lexerless parsing even more difficult...). pest is unapologetically pure PEG, but this makes parsing even really simple things that aren't easily described in a context-free manner near impossible. (Actually impossible, in the case of things that aren't formally context free, like Rust raw strings[^4].)

[^3]: Though perhaps too much so — I'm a fan of conditionally reserved words, and “soft keywords” are underused in grammar design. If a keyword can only show up in a position where identifiers cannot (e.g. it introduces an item and you don't allow defining items and non-items in the same syntactic context), it shouldn't be prohibited in identifiers.

[^4]: Which are technically context free, but only by the fact that the number of `#` has an upper bound, and the exact same logic proves *any* finite set of strings is itself context free, even if there is no better way to represent the set than pure enumeration.

TL;DR while I'm the defacto maintainer of the pest crates... I really dislike some of the fundamental choices pest makes. I could make a new parser that makes better choices... but while it may share syntax and a name, it wouldn't be pest.

Thanks for the ping, though; I'm going to make and promote a “reviving pest” discussions post to give org access to some community member willing to take over maintenance. Feel free to angrily @ me weekly if I don't get around to it 🙃

**Review by tomtau — COMMENTED** [review]
ok to merge this, but given there's no CI and I haven't tried building it, not sure if it still works.
@Jamalam360 was interested in reviving ast, so probably good to merge it?

**Comment by alex-ozdemir** [comment]
This patch still builds. Note, btw. That the transitive dependency failure has two security advisories [1] [2].

[1]: https://github.com/advisories/GHSA-r98r-j25q-rmpr
[2]: https://github.com/advisories/GHSA-jq66-xh47-j9f3
